Extend wallet restore endpoint to accept also account public key

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#1354 
#1355 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have updated swagger
- [x] I have added the endpoint and corresponding types
- [x] I have updated Api.TypesSpec
- [x] I have renamed `mkSeqState` to `mkSeqStateFromRootXPrv`
- [x] I have added `mkSeqStateFromAccountXPub` and property test
- [x] I have added impl of postAccountWallet
- [x] I have added integration tests illustrating the case


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
add endpoint types

complete adding endpoint types throught the code

update swagger and improve on json instances

small corrections

introduce AccountPostData and move everything under wallet

change swagger

use one POST type

improve on json instances of AccountPostData

remove unsafe from FromJSON

rename PostData

paweljakubas commented on lib/core/test/data/Cardano/Wallet/Api/HardwareWalletPostData.json:
```diff
@@ -0,0 +1,45 @@
+    "seed": -4365653956898380124,
+    "samples": [
+        {
+            "account_public_key": "6562643630396264366339313064643635653532306663393630306235353336353539623864363366653965313164613866633565366531633632626635653965343239343432613732653338613234383764613639393462613532376363376166636337366636323736643363393830366437333063363563646466636261",
```
I wonder if this account public key generated is realistic :thinking:

KtorZ replied:
The account public key is 64 bytes in length, so 128 characters hex-encoded.

paweljakubas replied:
indeed, I corrected the corresponding code

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -658,6 +674,22 @@ instance FromJSON WalletPostData where
+        case convertFromBase Base16 $ T.encodeUtf8 xs of
+            Left _ -> fail "wrong account public key format"
+            Right bs ->
+                pure $ AccountPublicKey $ ApiT $ unsafeDeserializeXPub bs
```
I am not sure if `unsafeDeserializeXPub` is ok here? The same for `serializeXPub` below

Anviking replied:
Will it not throw (i.e "something went wrong" for the user) if provided valid hex, but which is an invalid key?

KtorZ replied:
Whola! Yes. Do not use any `unsafe` method in the API when presented user-facing inputs :scream: ! 
The "unsafe" constructors are only tolerated at the boundaries with the database, where we have controlled first what we inserted!

paweljakubas replied:
it was temporary...now removed unsafe code. thanks for being vigilant here!

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -346,6 +355,20 @@ data ByronWalletPostData mw = ByronWalletPostData
+    } deriving (Eq, Generic, Show)
+
+newtype PostData = PostData
+    { postData :: Either WalletPostData AccountPostData
```
So the plan is to re-use the existing endpoint? Why is this preferable to having two endpoints?

When a wallet is created using just an account public key, will that not require special treatment inside the `cardano-wallet-core`, as it has no private key? Or does existing abstractions handle this already?

And if using the same endpoint, maybe call it `WalletOrAccountPostData` instead of `PostData`? Or `WalletPostData = SomeName WalletMnemonicPostData | SomeName2 WalletAccountPostData` or something 🤷‍♂  `PostData` sounds a bit too generic at least.

paweljakubas replied:
@Anviking yes, we will opt for ONE endpoint here. We talked about this with @KtorZ  and the conclusion was (when thinking about two endpoints): 
```
We are building a REST API, and we have to follow patterns and idioms of REST API. In REST, everything is about resources. Here, we are creating the same resource, so it should be the same endpoint.
```
Nevertheless, `WalletOrAccountPostData` is better than `PostData` - I agree

KtorZ replied:
> does existing abstractions handle this already?

Precisely. Attaching a private key is done as a separate operation, and, operation that requires a private key explicitly fails if there's none (e.g. signing tx)

> maybe call it WalletOrAccountPostData instead of PostData

:+1:

Anviking replied:
Btw, I find the use of `Either` (`postData :: Either WalletPostData AccountPostData`) _slightly_ weird because `Left` and `Right` imply success and error to me, and it has monad / functor instances. **But don't mind too much.**

I guess the tricky thing with `WalletOrAccountPostData = C1 WalletPostData | C2 AccountPostData` is naming the constructors. I'm not sure what conventions there are on it, but this is an example from `cardano-ledger`:

https://github.com/input-output-hk/cardano-ledger/blob/master/cardano-ledger/src/Cardano/Chain/Block/Block.hs#L359-L362

paweljakubas committed:
rename mkSeqState to mkSeqStateFromRootXPrv

paweljakubas committed:
add mkSeqStateFromAccountXPub with property test

paweljakubas committed:
add postAccountWallet impl

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -659,7 +680,7 @@ postShelleyWallet
     -> Handler ApiWallet
 postShelleyWallet ctx body = do
-    let state = mkSeqState (rootXPrv, pwd) g
+    let state = mkSeqStateFromRootXPrv (rootXPrv, pwd) g
```
👍

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:
```diff
@@ -450,8 +451,11 @@ instance PersistPublicKey (ShelleyKey depth) where
-        xpubFromText = xpub <=< fromHex @ByteString
         err _ = error "unsafeDeserializeXPub: unable to deserialize ShelleyKey"
+
+xpubFromText :: ByteString -> Either String XPub
```
But we lose the ability to distinguish between "invalid hex" and "correct hex, invalid data" from the `FromJSON` because of this function?

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -571,6 +572,27 @@ mkSeqState (rootXPrv, pwd) g =
+mkSeqStateFromAccountXPub accXPub g =
+    let
+        rewardXPub =
+            deriveAddressPublicKey accXPub MutableAccount minBound
```
So this matches the reward address for "normal wallets", right? Assuming the accountXPub is the first account.

Here `minBound` is _also_ the first _Soft_ index  https://github.com/input-output-hk/cardano-wallet/blob/b9daac5b448308cd807163028af4b90d71f22352/lib/core/src/Cardano/Wallet.hs#L1167

because of 
```haskell
instance HardDerivation ShelleyKey where
    type AddressIndexDerivationType ShelleyKey = 'Soft
```
(so it should work. _Wasn't obvious to me at first_)

KtorZ replied:
>  Wasn't obvious to me at first

This indeed deserve a proper note as a comment in the source code. All your assumptions are correct, but we need to explain _why_ they are and what's the rationale behind this.

Anviking replied:
Another gotcha I realised was that:
```haskell
deriveAddressPrivateKey _ _ _ minBound
```
infers `minBound` to be the first **soft** index (for ShelleyKey)

but for
```haskell
deriveAccountPrivateKey _ _ _ minBound
```
`minBound` is the first **hardened** index (regardless of `AddressIndexDerivationType`). https://github.com/input-output-hk/cardano-wallet/blob/b9daac5b448308cd807163028af4b90d71f22352/lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L285

(hence the `0x80000000` index in the integration test)

_type-inference…_ 😅

paweljakubas replied:
added the suggested comment

Anviking commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1035,6 +1045,22 @@ instance Arbitrary SortOrder where
+        let walletPostDataGen = arbitrary :: Gen WalletPostData
+        let accountPostDataGen = arbitrary :: Gen AccountPostData
+        oneof [ WalletOrAccountPostData . Left <$> walletPostDataGen
+              , WalletOrAccountPostData . Right <$> accountPostDataGen ]
```
Now, there's no shrinking. Might not be _that_ much of a problem…?

But is there any specific reason not to use `genericArbitrary` and `genericShrink` for `WalletOrAccountPostData`, and `AccountPostData` too?

paweljakubas replied:
We are going to have either one or other object - I think it is ok to see them as they fail

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs:
```diff
@@ -335,16 +339,32 @@ prop_poolEventuallyDiscoverOurs _proxy (g, addr) =
+
+-- | We can always generate at exactly `gap` change addresses (on the internal
+-- chain) using mkSeqStateFromAccountXPub
+prop_genChangeGapFromAccountXPub
```
Is this really needed?

Is this not more a property of `SeqState` than of `mkSeqStateFromRootXPrv`/`mkSeqStateFromAccountXPub `?

paweljakubas replied:
I just treat it as some place to check that both `mkSeqStateFromRootXPrv` and `mkSeqStateFromAccountXPub` produce consistent results. The test is not utterly persuasive but gives some value added (in my modest opinion) ;-)

Anviking commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -301,6 +309,8 @@ spec = do
             jsonRoundtripAndGolden $ Proxy @(PostTransactionFeeData 'Testnet)
             jsonRoundtripAndGolden $ Proxy @WalletPostData
+            jsonRoundtripAndGolden $ Proxy @AccountPostData
+            jsonRoundtripAndGolden $ Proxy @WalletOrAccountPostData
```
👍

paweljakubas committed:
add integration test

fix passwd

remove magic number and use minBound

Anviking reviewed (approved):
Lgtm.
Maybe most significantly (see comments):
- is `prop_genChangeGapFromAccountXPub` really needed?
- Should we not be able to distinguish between invalid hex encoding and invalid data?

paweljakubas commented on specifications/api/swagger.yaml:
```diff
@@ -845,6 +853,20 @@ definitions:
+      account_public_key: *walletAccountPubkey
+      address_pool_gap: *walletAddressPoolGap
+
+  ApiWalletOrAccountPostData: &ApiWalletOrAccountPostData
```
so I have problem here. how to express `oneof` . This one unfortunately gives : 
```
 1) Cardano.Wallet.Api.Types, verify that every type used with JSON content type in a servant API has compatible ToJSON and ToSchema instances using validateToJSON., WalletOrAccountPostData
       Falsifiable (after 3 tests):
         WalletOrAccountPostData {postData = Right (AccountPostData {name = ApiT {getApiT = WalletName {getWalletName = ":p6&vFta~? Yp\147454n\30864rWs~\"!r#:ms 3+KZUelcepj>n$0F99`$UVGunr\27692}V0xnw3+0CJ_fe/\152955N|)EBVO6Yd[u%,(rSJ:"}}, accountPublicKey = AccountPublicKey {key = ApiT {getApiT = ShelleyKey {getKey = XPub {xpubPublicKey = "Qd\237Bm\207\154\179\ESC>\208I\219\236c|\b\174\141\199w\179\SUB\242\183\RS\219\186\&7\179\&5\194", xpubChaincode = ChainCode "\226c\ENQP\DC3#\FS\"\221\133\150n@[33\154\138xh\133%\187V\181^6,v\220\131l"}}}}, addressPoolGap = Nothing})}
         Validation against the schema fails:
           * property "mnemonic_sentence" is required, but not found in "{\"account_public_key\":\"5164ed426dcf9ab31b3ed049dbec637c08ae8dc777b31af2b71edbba37b335c2e263055013231c22dd85966e405b33339a8a78688525bb56b55e362c76dc836c\",\"name\":\":p6&vFta~? Yp\240\163\191\190n\231\162\144rWs~\\\"!r#:ms 3+KZUelcepj>n$0F99`$UVGunr\230\176\172}V0xnw3+0CJ_fe/\240\165\149\187N|)EBVO6Yd[u%,(rSJ:\"}"
           * property "passphrase" is required, but not found in "{\"account_public_key\":\"5164ed426dcf9ab31b3ed049dbec637c08ae8dc777b31af2b71edbba37b335c2e263055013231c22dd85966e405b33339a8a78688525bb56b55e362c76dc836c\",\"name\":\":p6&vFta~? Yp\240\163\191\190n\231\162\144rWs~\\\"!r#:ms 3+KZUelcepj>n$0F99`$UVGunr\230\176\172}V0xnw3+0CJ_fe/\240\165\149\187N|)EBVO6Yd[u%,(rSJ:\"}"
           * property "account_public_key" is found in JSON value, but it is not mentioned in Swagger schema
         
         JSON value:
         {
             "account_public_key": "5164ed426dcf9ab31b3ed049dbec637c08ae8dc777b31af2b71edbba37b335c2e263055013231c22dd85966e405b33339a8a78688525bb56b55e362c76dc836c",
             "name": ":p6&vFta~? Yp𣿾n碐rWs~\"!r#:ms 3+KZUelcepj>n$0F99`$UVGunr氬}V0xnw3+0CJ_fe/𥕻N|)EBVO6Yd[u%,(rSJ:"
         }
         
         Swagger Schema:
         {
             "required": [
                 "name",
                 "mnemonic_sentence",
                 "passphrase"
             ],

```

paweljakubas replied:
fixed in https://github.com/input-output-hk/cardano-wallet/pull/1360/commits/794224a42c9a7f7e646bdd58a76572f670c1260c

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -658,6 +681,39 @@ instance FromJSON WalletPostData where
+                pure $ WalletOrAccountPostData $ Right xs
+            Just _ -> do
+                xs <- parseJSON obj :: Aeson.Parser WalletPostData
+                pure $ WalletOrAccountPostData $ Left xs
```
Tips: parsers are alternatives. So we can simply write:

```hs
parseJSON obj = WalletOrAccountPostData <$>
    (Left <$> parseJSON obj <|> Right <$> parseJSON obj)
```

;)

Anviking replied:
Could it not potentially result in worse error messages?

If there is a slight mistake in the `Left` branch, will it not continue with the `Right` branch and might produce an error message specific to the right branch?

KtorZ replied:
That's pretty much the case now too. If you forget the `passphrase` (or mistype it), we'll just assume an `AccountPostData` :man_shrugging:

We can probably do something more elaborate like we did for `FromMnemonic`. My two cents would actually be that, we should probably collect the error message from both branches and provide both to the user in a readable way.

paweljakubas replied:
done

paweljakubas replied:
hmmm, I have a number of things like that : 
```

1) API Specifications, WALLETS_CREATE_05 - Mnemonics, [] as mnemonic_sentence -> fail
--
  | "Error in $: key \"id\" not present: Response {responseStatus = Status {statusCode = 400, statusMessage = \"Bad Request\"}, responseVersion = HTTP/1.1, responseHeaders = [(\"Transfer-Encoding\",\"chunked\"),(\"Date\",\"Mon, 24 Feb 2020 19:49:06 GMT\"),(\"Server\",\"Warp/3.2.27\"),(\"Content-Type\",\"application/json;charset=utf-8\")], responseBody = \"{\\\"code\\\":\\\"bad_request\\\",\\\"message\\\":\\\"Error in $: key 'account_public_key' not present\\\"}\", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}" does not contain "Invalid number of words: 15, 18, 21 or 24 words are expected."
```
will need to work on error management in this instance

paweljakubas committed:
try use oneof

fixing oneOf

add type to ApiWalletOrAccountPostData

improve on description of swagger.yaml

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1470,6 +1495,12 @@ instance ToSchema ApiTxId where
+    declareNamedSchema _ = declareSchemaForDefinition "ApiAccountPostData"
+
+instance ToSchema WalletOrAccountPostData where
+    declareNamedSchema _ = declareSchemaForDefinition "ApiWalletOrAccountPostData"
```
will need probably to do something here, utilize `proxy` in `declareNamedSchema proxy` and basing on that in one case 
```
declareSchemaForDefinition "ApiAccountPostData"
```
and 
```
declareSchemaForDefinition "ApiWalletPostData"
```

paweljakubas replied:
fixed in 794224a

KtorZ committed:
Work around 'oneOf' not being supported in swagger2 for validateEveryToJSON

paweljakubas committed:
final wrap up after reviews

improve FromJSON WalletOrAccountPostData

rebase fix to new eventually
